kafka-client/src/main/java/dev/responsive/kafka/internal/stores/CommitBuffer.java

```diff
@@ -408,23 +412,51 @@ private RemoteWriteResult drain(final Collection<RemoteWriter<K>> writers) {
+   * in the background. After a flush, we will check on the completion of the previous
+   * delete request and if successful, send out a new request to truncate up to the
+   * last flushed offset
+   *
+   * @param offset the flushed offset up to which we can safely delete changelog records
+   */
   private void maybeTruncateChangelog(final long offset) {
-    if (truncateChangelog && !isNotDeleteEnabled) {
-      try {
-        admin.deleteRecords(Map.of(changelog, RecordsToDelete.beforeOffset(offset))).all().get();
-        log.info("Truncated changelog topic {} before offset {}", changelog, offset);
-      } catch (final ExecutionException e) {
-        log.warn("Could not truncate changelog topic-partition " + changelog, e);
-        if (e.getCause() instanceof PolicyViolationException) {
-          log.warn("Disabling further changelog truncation attempts due to topic configuration "
-                       + "being incompatible with deleteRecord requests", e);
-          isNotDeleteEnabled = true;
+    if (deleteRecordsResult != null) {
+      final KafkaFuture<DeletedRecords> future = deleteRecordsResult.lowWatermarks().get(changelog);
+
+      if (future.isDone()) {
+
+        if (future.isCancelled()) {
+          log.warn("Delete records request for changelog {} was unexpectedly cancelled, "
+                       + "will retry the topic truncation", changelog);
+        } else {
+          try {
+            final DeletedRecords deleted = future.get();
+            log.info("Truncated changelog {} up to offset {}", changelog, deleted.lowWatermark());
+          } catch (final ExecutionException e) {
+            if (e.getCause() instanceof PolicyViolationException) {
+              log.warn("Disabling further changelog truncation attempts due to topic configuration "
+                           + "being incompatible with deleteRecords requests", e);
+              isNotDeleteEnabled = true;
+            } else {
+              log.warn("Truncation of changelog " + changelog + " failed and will be retried", e);
+            }
+          } catch (final InterruptedException e) {
+            log.error("Interrupted while truncating the changelog " + changelog, e);
+            throw new RuntimeException(logPrefix + "Interrupted while truncating " + changelog, e);
+          }
         }
-      } catch (final InterruptedException e) {
-        log.error("Interrupted while truncating the changelog " + changelog, e);
-        throw new ProcessorStateException("Interrupted while truncating " + changelog, e);
+        deleteRecordsResult = null;
+
+      } else {
+        log.debug("Still waiting on previous changelog truncation attempt to complete");
       }
     }
+
+    if (truncateChangelog && !isNotDeleteEnabled && deleteRecordsResult == null) {
+      log.debug("Issuing new delete request to truncate {} up to offset {}", changelog, offset);
+      deleteRecordsResult = admin.deleteRecords(Map.of(changelog, beforeOffset(offset)));
```

Comment:
(feel free to ignore) if you're keen to explore, I think this would be more idiomatic if you change this to use `Future#whenComplete` and pass in the logging as a method reference

then you can also initialize it to `KafkaFuture.completedFuture(null)` and avoid nulls by:
```
if (truncateChangelog && isNotDeleteEnabled && deleteRecordResult.isDone()) {
   deleteRecordResult = admin.deleteRecords().lowWatemarks().get(changelog);
}
```

Reply:
ah yeah that's much nicer, good tip